treewide: rename fetchzip -> fetchArchive

DavHau opened the pull request:
After 2 years of using nix, I finally learn that there actually is a nixpkgs alternative for `builtins.fetchTarball`.
This is way too late and the root cause is a naming issue which I want to fix with this PR.

Done:
  - add alias `pkgs.fetcharchive` for `pkgs.fetchzip`

bjornfor commented:
> we could also name it `fetchArchive` instead

Observation: that naming style is more common with nix builtins (fetchTarball, fetchGit) than nixpkgs (fetchurl, fetchgit).

DavHau commented:
Can this be merged? If so, I'm happy to rebase it once more.

roberth commented:
Please separate this into two commits
1. fetchArchive itself, changes to the docs, etc
2. the treewide commit, performing the rename at all call sites.

roberth commented:
Please separate this into two commits
1. fetchArchive itself, changes to the docs, etc
2. the treewide commit, performing the rename at all call sites.

Artturin commented:
>     1. fetchArchive itself, changes to the docs, etc

+ a release notes

DavHau committed:
fetchzip: rename to fetchArchive

- this clarifies this fetcher can handle archives other than .zip
- for backwards compatibility `fetchzip` is kept as an alias

DavHau committed:
treewide: rename fetchzip -> fetchArchive

Following the renaming of the previous commit

DavHau commented:
@roberth @Artturin I applied your requests.

sternenseemann commented:
I'm sceptical that this is worth it. Sure, naming is suboptimal, but this is just gonna break backwards compatibility for a small benefit. The new `fetchzip` isn't even supposed to be used for all archives, since the whole situation in nixpkgs is relatively subtle and thus confusing, e. g. fetching an archive with `fetchurl` and relying on `unpackPhase`.

DavHau commented:
@sternenseemann There is still an alias `fetchzip`, therefore it is not breaking backwards compatiblity.

sternenseemann commented:
An alias is always a road to deprecation and removal: I don't think you propose carrying this indefinitely. So the backwards incompatible change needs to be discussed now.

Also it is one when evaluating with `config.allowAliases = false`.

DavHau commented:
I wouldn't mind carrying that alias indefinitely. Checking the git blame of the aliases.nix file, It looks like the majority of aliases are several years old. It seems to be normal for aliases to stay for a long time or even forever.

I think most people agree, that the nix community has a big usability problem which is holding adoption back.
Inconsistent and weird naming of stuff contributes to that and therefore I think fixing it is important.

The name `fetchzip` is particularly bad, as the first impression gives me a very clear idea about what it does, so I wouldn't even consider reading the manual, which keeps me from learning that the first impression is just plain wrong.
Wrong names are worse than weird names.

roberth commented:
> Also it is one when evaluating with `config.allowAliases = false`.

This is an undocumented config item, intended for internal CI. If you use it in your own systems and projects, that is a personal choice; not one of the general user base. _That said, churn does have a cost,_ because users want to be warning-clean. This puts a burden on maintainers, regardless of how simple a change may be.
I have failed to recognize this recently when reviewing the `literalExpression` change and I still feel bad about it, but it was too late to make a final improvement that would have helped downstream maintainers. See, while the change looked simple, it wasn't for the maintainers, because the "new way" wasn't available in older versions that they still needed to support.

So what's missing in our current approach to such small changes is a waiting period; waiting until _all_ supported versions of Nixpkgs support the new way, before even introducing a warning, because the warning will make users trigger the maintainers too soon.

SuperSamus commented on doc/builders/fetchers.chapter.md:
````diff
@@ -24,7 +24,7 @@ stdenv.mkDerivation {
 ```
 
-The main difference between `fetchurl` and `fetchzip` is in how they store the contents. `fetchurl` will store the unaltered contents of the URL within the Nix store. `fetchzip` on the other hand will decompress the archive for you, making files and directories directly accessible in the future. `fetchzip` can only be used with archives. Despite the name, `fetchzip` is not limited to .zip files a ...
+The main difference between `fetchurl` and `fetchArchive` is in how they store the contents. `fetchurl` will store the unaltered contents of the URL within the Nix store. `fetchArchive` on the other hand will decompress the archive for you, making files and directories directly accessible in the future. `fetchArchive` can only be used with archives. Despite the name, `fetchArchive` is not limited ...
````
The last part of this paragraph can be removed.
